src/main/java/org/folio/rest/impl/InvoiceLineHelper.java

```diff
@@ -480,4 +490,64 @@ private String getInvoiceLineNumberEndpoint(String id) {
 
   }
 
+  /**
+   * Updates the invoice's poNumbers field, following an invoice line creation, update or removal.
+   * @param invoice - the invoice of the modified invoice line
+   * @param invoiceLine - the modified invoice line
+   * @param invoiceLineFromStorage - the old version of the invoice line
+   * @param requestContext - used to start new requests
+   */
+  private CompletableFuture<Void> updateInvoicePoNumbers(Invoice invoice, InvoiceLine invoiceLine,
+      InvoiceLine invoiceLineFromStorage, RequestContext requestContext) {
+
+    if (invoiceLineFromStorage == null && invoiceLine.getPoLineId() == null)
+      return CompletableFuture.completedFuture(null);
+    if (invoiceLine.getPoLineId() == null && invoiceLineFromStorage != null && invoiceLineFromStorage.getPoLineId() == null)
+      return CompletableFuture.completedFuture(null);
+    String poLineId = (invoiceLineFromStorage == null || invoiceLine.getPoLineId() != null) ? invoiceLine.getPoLineId() :
+      invoiceLineFromStorage.getPoLineId();
+    return orderService.getPoLine(poLineId, requestContext)
+      .thenCompose(poLine -> orderService.getOrder(poLine.getPurchaseOrderId(), requestContext))
+      .thenCompose(order -> {
+        List<String> invoicePoNumbers = invoice.getPoNumbers();
+        String orderPoNumber = order.getPoNumber();
+        if (orderPoNumber == null)
+          return CompletableFuture.completedFuture(null);
+        if (invoiceLineFromStorage != null && invoiceLineFromStorage.getPoLineId() != null && invoiceLine.getPoLineId() == null) {
+          // remove the po number if needed
+          if (invoicePoNumbers == null || !invoicePoNumbers.contains(orderPoNumber))
+            return CompletableFuture.completedFuture(null);
+          // check the other invoice lines to see if one of them is linking to the same order
+          List<String> orderLineIds = order.getCompositePoLines().stream().map(CompositePoLine::getId).collect(toList());
+          return getRelatedLines(invoiceLine).thenCompose(lines -> {
+            if (lines.stream().anyMatch(line -> orderLineIds.contains(line.getPoLineId()))) {
+              return CompletableFuture.completedFuture(null);
+            } else {
+              List<String> newNumbers = invoicePoNumbers.stream().filter(n -> !n.equals(orderPoNumber)).collect(toList());
+              return invoiceService.updateInvoice(invoice.withPoNumbers(newNumbers), requestContext);
+            }
+          });
+        }
+        // add the po number if needed
+        if (invoicePoNumbers != null && invoicePoNumbers.contains(orderPoNumber))
```

Comment:
the same redundant check. ObjectMapper always initializes lists as `new ArrayList<>()`